lib/ds/arrayutils.flow

```diff
@@ -518,20 +518,31 @@ foldiM(xs : [?], init : ??, fn : (int, ??, ?) -> Maybe<??>) -> Maybe<??> {
 	foldi(xs, Some(init), \i, accM : Maybe<??>, x : ? -> maybeBind(accM, \acc : ?? -> fn(i, acc, x)));
 }
 
-splitModulo(a : [?], m : int) -> [[?]] {
-	n = length(a) / m;
-	rest = length(a) % m;
-	generate(0, m, \i ->
-		generate(0, if (i < rest) n + 1 else n, \j ->
-			a[j * m + i]
-		)
-	);
+splitModulo(a : [?], m0 : int) -> [[?]] {
+	m = iabs(m0);
+	if (m == 0) map(a, \x -> [x]) else {
+		n = length(a) / m;
+		rest = length(a) % m;
+		generate(0, m, \i ->
+			generate(0, if (i < rest) n + 1 else n, \j ->
+				a[j * m + i]
+			)
+		);
+	}
 }
 
 concatAModulo(a : [[?]]) -> [?] {
 	m = length(a);
 	len = fold(a, 0, \acc, x -> acc + length(x));
-	generate(0, len, \i ->
-		a[i % m][i / m]
-	);
+	j = ref 0;
+	i = ref for(0, \k -> (k < m) && (^j >= length(a[k])), \k -> k + 1);
+	generate(0, len, \__ -> {
+		x = a[^i][^j];
```

Comment:
unsafe retrieval of element by index

Reply:
Here retrieval is safe. It may be proved.